connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java

```diff
@@ -0,0 +1,524 @@
+
+import org.apache.kafka.common.utils.AbstractIterator;
+import org.apache.kafka.connect.data.Date;
+import org.apache.kafka.connect.data.Decimal;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.data.Time;
+import org.apache.kafka.connect.data.Timestamp;
+import org.apache.kafka.connect.errors.DataException;
+
+import java.math.BigDecimal;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A basic {@link Headers} implementation.
+ */
+public class ConnectHeaders implements Headers {
+
+    private static final int EMPTY_HASH = Objects.hash(new LinkedList<>());
+
+    // Shared, immutable instance with a null schema and null value
+    private static final SchemaAndValue NULL_SCHEMA_AND_VALUE = new SchemaAndValue(null, null);
+
+    /**
+     * An immutable and therefore sharable empty iterator.
+     */
+    private static final Iterator<Header> EMPTY_ITERATOR = new Iterator<Header>() {
+        @Override
+        public boolean hasNext() {
+            return false;
+        }
+
+        @Override
+        public Header next() {
+            throw new NoSuchElementException();
+        }
+
+        @Override
+        public void remove() {
+            throw new IllegalStateException();
+        }
+    };
+
+
+    // This field is set lazily, but once set to a list it is never set back to null
+    private volatile LinkedList<Header> headers;
+
+    public ConnectHeaders() {
+    }
+
+    public ConnectHeaders(Iterable<Header> original) {
+        if (original != null) {
+            if (original instanceof ConnectHeaders) {
+                ConnectHeaders headers = (ConnectHeaders) original;
+                if (!headers.isEmpty()) {
+                    this.headers = new LinkedList<>(((ConnectHeaders) original).headers);
+                }
+            } else {
+                this.headers = new LinkedList<>();
+                for (Header header : original) {
+                    this.headers.add(header);
+                }
+            }
+        }
+    }
+
+    @Override
+    public int size() {
+        return this.headers == null ? 0 : headers.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return this.headers == null ? true : headers.isEmpty();
+    }
+
+    @Override
+    public Headers clear() {
+        if (headers != null) {
+            headers.clear();
+        }
+        return this;
+    }
+
+    @Override
+    public Headers add(Header header) {
+        Objects.requireNonNull(header, "Unable to add a null header.");
+        if (this.headers == null) {
+            synchronized (this) {
```

Comment:
The whole point of this is to not actually create the list if there are no headers on a record. I'd expect that to be by far the prevalent case, so I thought it worthwhile to deal with that.

However, as you suggest, we should only have single threaded access, so the use of `volatile` and double-checked locking is, to say the least, ridiculous.

Reply:
I get the intention for lazy initialization. Asked generally since I wanted to be 100% sure about the expectations w.r.t. concurrency. 

I think it's best to keep `volatile` since the field is not final and might be accessed in sequence by different threads, but `synchronized` can be skipped if there's no concurrent execution by different threads.